Source/drlg_l2.cpp

```diff
@@ -1723,50 +1727,40 @@ bool PlaceMiniSet(const BYTE *miniset, int tmin, int tmax, int cx, int cy, bool
 	return true;
 }
 
-void PlaceMiniSetRandom(const BYTE *miniset, int rndper)
+void PlaceMiniSetRandom(const Miniset &miniset, int rndper)
 {
-	int sw = miniset[0];
-	int sh = miniset[1];
+	int sw = miniset.size.width;
+	int sh = miniset.size.height;
 
 	for (int sy = 0; sy < DMAXY - sh; sy++) {
 		for (int sx = 0; sx < DMAXX - sw; sx++) {
-			bool found = true;
-			int ii = 2;
-			if (sx >= nSx1 && sx <= nSx2 && sy >= nSy1 && sy <= nSy2) {
-				found = false;
-			}
-			for (int yy = 0; yy < sh && found; yy++) {
-				for (int xx = 0; xx < sw && found; xx++) {
-					if (miniset[ii] != 0 && dungeon[xx + sx][yy + sy] != miniset[ii]) {
-						found = false;
-					}
-					if (dflags[xx + sx][yy + sy] != 0) {
-						found = false;
-					}
-					ii++;
+			if (sx >= nSx1 && sx <= nSx2 && sy >= nSy1 && sy <= nSy2)
+				goto next;
+			for (int yy = 0; yy < sh; yy++) {
+				for (int xx = 0; xx < sw; xx++) {
+					if (miniset.search[yy][xx] != 0 && dungeon[xx + sx][yy + sy] != miniset.search[yy][xx])
+						goto next;
+					if (dflags[xx + sx][yy + sy] != 0)
+						goto next;
 				}
 			}
-			int kk = sw * sh + 2;
-			if (found) {
-				for (int yy = std::max(sy - sh, 0); yy < std::min(sy + 2 * sh, DMAXY) && found; yy++) {
-					for (int xx = std::max(sx - sw, 0); xx < std::min(sx + 2 * sw, DMAXX); xx++) {
-						// BUGFIX: yy and xx can go out of bounds (fixed)
-						if (dungeon[xx][yy] == miniset[kk]) {
-							found = false;
-						}
-					}
+			for (int yy = std::max(sy - sh, 0); yy < std::min(sy + 2 * sh, DMAXY); yy++) {
+				for (int xx = std::max(sx - sw, 0); xx < std::min(sx + 2 * sw, DMAXX); xx++) {
+					// BUGFIX: yy and xx can go out of bounds (fixed)
+					if (dungeon[xx][yy] == miniset.replace[0][0])
+						goto next;
 				}
 			}
-			if (found && GenerateRnd(100) < rndper) {
-				for (int yy = 0; yy < sh; yy++) {
-					for (int xx = 0; xx < sw; xx++) {
-						if (miniset[kk] != 0) {
-							dungeon[xx + sx][yy + sy] = miniset[kk];
-						}
-						kk++;
-					}
+			if (GenerateRnd(100) >= rndper)
+				goto next;
+			for (int yy = 0; yy < sh; yy++) {
+				for (int xx = 0; xx < sw; xx++) {
+					if (miniset.search[yy][xx] != 0)

```

Comment:
```suggestion
					if (miniset.replace[yy][xx] != 0)
```
The old logic was using kk (the replace index) for the check and assignment